### PR TITLE
Bug fix: Make collector not fail on null values

### PR DESCRIPTION
**Pull request by Sticksman**
Resolves: https://github.com/prometheus-community/postgres_exporter/issues/818

Make all row scans that gather values nullable except the ones that gather absolute values. Nulls are emitted to prometheus as "0" values or "unknown" for labels. Also wrote an additional set of tests to make sure nulls are caught and dealt with properly.

### CODE REVIEW

File: collector/pg_postmaster.go [hunk 1]
```diff
@@ -51,14 +52,20 @@ func (c *PGPostmasterCollector) Update(ctx context.Context, instance *instance,
...
 	err := row.Scan(&startTimeSeconds)
 	if err != nil {
 		return err
 	}
+	var startTimeSecondsMetric float64
+	if startTimeSeconds.Valid {
+		startTimeSecondsMetric = startTimeSeconds.Float64
+	} else {
+		startTimeSecondsMetric = 0
+	}
```
Comment (SuperQ) [inline collector/pg_postmaster.go:65]:
```suggestion
	startTimeSecondsMetric := 0.0
	if startTimeSeconds.Valid {
		startTimeSecondsMetric = startTimeSeconds.Float64
	}
```

File: collector/pg_process_idle.go [hunk 2]
```diff
@@ -97,10 +99,26 @@ func (PGProcessIdleCollector) Update(ctx context.Context, instance *instance, ch
 	if err != nil {
 		return err
 	}
+
+	var applicationNameLabel string
+	if applicationName.Valid {
+		applicationNameLabel = applicationName.String
+	} else {
+		applicationNameLabel = "unknown"
+	}
```
Comment (SuperQ) [inline collector/pg_process_idle.go:108]:
```suggestion
	applicationNameLabel := "unknown"
	if applicationName.Valid {
		applicationNameLabel = applicationName.String
	}
```

**Review by SuperQ — COMMENTED** [review]
Minor Go style nit.

File: collector/pg_database.go [hunk 3]
```diff
@@ -79,32 +80,41 @@ func (c PGDatabaseCollector) Update(ctx context.Context, instance *instance, ch
...
 		if err != nil {
 			return err
 		}
 
+		var sizeMetric float64
+		if size.Valid {
+			sizeMetric = size.Float64
+		} else {
+			sizeMetric = 0
+		}
```
Comment (SuperQ) [inline collector/pg_database.go:114]:
```suggestion
		sizeMetric := 0.0
		if size.Valid {
			sizeMetric = size.Float64
		}
```

File: collector/pg_replication_slot.go [hunk 4]
```diff
@@ -82,32 +83,44 @@ func (PGReplicationSlotCollector) Update(ctx context.Context, instance *instance
...
+		var flushLSN sql.NullFloat64
+		var isActive sql.NullBool
 		if err := rows.Scan(&slotName, &walLSN, &flushLSN, &isActive); err != nil {
 			return err
 		}
 
 		isActiveValue := 0
-		if isActive {
+		if isActive.Valid && isActive.Bool {
 			isActiveValue = 1
```
Comment (SuperQ) [inline collector/pg_replication_slot.go:96]: Can simplify to remove need for `float64()` later:

```go
		isActiveValue := 0.0
		if isActive.Valid && isActive.Bool {
			isActiveValue = 1.0
```

File: collector/pg_stat_database.go [hunk 5]
```diff
@@ -280,152 +267,218 @@ func (PGStatDatabaseCollector) Update(ctx context.Context, instance *instance, c
...
+		datidLabel := "unknown"
+		if datid.Valid {
+			datidLabel = datid.String
+		}
+		datnameLabel := "unknown"
+		if datname.Valid {
+			datnameLabel = datname.String
+		}
 
+		var numBackendsMetric float64
```
Comment (SuperQ) [inline collector/pg_stat_database.go:279]: This should probably be reset to 0 to avoid row-torow leaking int the for loop.

```suggestion
		numBackendsMetric := 0.0
```

File: collector/pg_stat_database_test.go [hunk 6]
```diff
@@ -0,0 +1,220 @@
...
+			2034563757,
+			0,
+			2725688749,
+			23,
+			52,
+			74,
+			925,
+			16,
+			823,
+			srT)
```
Comment (SuperQ) [inline collector/pg_stat_database_test.go:82]: Can we add a second row with all 0 entries to make sure we don't have any row-to-row data leaks?

**Comment by Sticksman** [comment]
Made all `var foo float64` declarations into `foo := 0.0`. Added a row leak test, but I think it should be fine, considering lint yells at us about `var foo type` being a 0 declaration

**Review by SuperQ — APPROVED** [review]
Nice

File: collector/pg_database.go [hunk 7]
```diff
@@ -79,32 +80,39 @@ func (c PGDatabaseCollector) Update(ctx context.Context, instance *instance, ch
 	var databases []string
 
 	for rows.Next() {
-		var datname string
+		var datname sql.NullString
```
Comment (sysadmind) [inline collector/pg_database.go:83]: Is this column nullable? The documentation doesn't mention that this field is nullable.

File: collector/pg_database.go [hunk 8]
```diff
@@ -79,32 +80,39 @@ func (c PGDatabaseCollector) Update(ctx context.Context, instance *instance, ch
...
 		}
 
-		databases = append(databases, datname)
+		databases = append(databases, datname.String)
 	}
 
 	// Query the size of the databases
 	for _, datname := range databases {
-		var size int64
+		var size sql.NullFloat64
```
Comment (sysadmind) [inline collector/pg_database.go:103]: I also don't see any documentation that says that this field can be NULL>

**Review by sysadmind — CHANGES_REQUESTED** [review]
The more that I look at this PR, the more that it appears it is blindly using nullable data types for every query. I don't think that's the best approach. I think that the data types should change only where the field can actually be null.

**Comment by SuperQ** [comment]
@sysadmind I think we should handle nullable for basically everything, since you never know when PG will either change the underlying data between versions, or have incorrect documentation. :-/

**Comment by Sticksman** [comment]
datname shouldn't be nullable, but one the original report included it as a field that returned null: https://github.com/prometheus-community/postgres_exporter/issues/818#issuecomment-1604358077

File: collector/pg_process_idle.go [hunk 9]
```diff
@@ -15,21 +15,23 @@ package collector
...
 	"github.com/go-kit/log"
 	"github.com/prometheus/client_golang/prometheus"
 )
 
-const processIdleSubsystem = "process_idle"
-
 func init() {
-	registerCollector(processIdleSubsystem, defaultEnabled, NewPGProcessIdleCollector)
+	// Making this default disabled because we have no tests for it
+	registerCollector("statements", defaultDisabled, NewPGProcessIdleCollector)
```
Comment (sysadmind) [inline collector/pg_process_idle.go:26]: Why was this renamed?

**Review by sysadmind — COMMENTED** [review]
One question this round. 

@SuperQ I'm not sure that I agree that everything should be NULLable. This adds boilerplate for every field and makes the code harder to read. If you feel that this is the best path, I won't protest but I don't love it.

**Comment by SuperQ** [comment]
I don't love it either, but I don't know enough about PostgreSQL to know if we should one-off fix NULL errors or deal with the boilerplate.

File: collector/pg_process_idle.go (hunk 9 above)
Comment (Sticksman) [reply collector/pg_process_idle.go:26]: Good question. Probably a typo somewhere

File: collector/pg_stat_database_test.go (hunk 6 above)
Comment (sysadmind) [reply collector/pg_stat_database_test.go:82]: This shows as resolved, but I don't see any of these with 2 rows

Comment (SuperQ) [reply collector/pg_stat_database_test.go:82]: Oh, I saw some second rows in other tests.

Comment (SuperQ) [reply collector/pg_stat_database_test.go:82]: Oh, yea, see `TestPGStatDatabaseCollectorRowLeakTest` below.

**Review by sysadmind — APPROVED** [review]
I don't think that this is as clean as I would like but I think it is at a state that we can iterate from and at least resolve the bugs for our users.

**Review by SuperQ — APPROVED** [review]
